src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -74,189 +58,331 @@ class TaskReplaceActor(
+        context.become(killing)
+        self ! KillNext
+      }
+
+    // Stash all instance changed events
     case stashMe: AnyRef =>
       stash()
   }
 
-  private def initialized: Receive = readinessBehavior orElse replaceBehavior
-
-  /**
-    * This actor is a bad example on how future orchestrator might handle the instances. The logic below handles instance
-    * changed across three dimensions:
-    *
-    * a) old vs. new - instances with old RunSpec version are gradually replaced with the new one
-    * b) goals - it's not enough to check instance condition e.g. if the new instance task FAILED but the goal is
-    *            Goal.Running then it will be automatically rescheduled by the TaskLauncherActor
-    * c) condition - we additionally check whether or not the instance is considered terminal/active
-    *
-    * What makes it so hard to work with, is the fact, that it basically counts old and new instances and the additional
-    * dimensions are expressed through filters on the [[InstanceChanged]] events. It can be a more robust state-machine
-    * which ideally has a set of new and old instances which then decommissions old ones and schedules new ones, never
-    * incrementing/decrementing counters and never over/under scales.
-    *
-    */
-  def replaceBehavior: Receive = {
-
-    // === An InstanceChanged event for the *new* instance ===
-    case ic: InstanceChanged if !oldInstanceIds(ic.id) =>
-      val id = ic.id
-      val condition = ic.condition
-      val instance = ic.instance
-      val goal = instance.state.goal
-      val agentId = instance.agentInfo.fold(Option.empty[String])(_.agentId)
-
-      // 1) Did the new instance task fail?
-      if (considerTerminal(condition) && goal == Goal.Running) {
-        logger.warn(s"New $id is terminal ($condition) on agent $agentId during app $pathId restart: $condition reservation: ${instance.reservation}. Waiting for the task to restart...")
-        instanceTerminated(id)
-        instancesStarted -= 1
-      } // 2) Did someone tamper with new instance's goal? Don't do that - there should be only one "orchestrator" per service per time!
-      else if (considerTerminal(condition) && goal != Goal.Running) {
-        logger.error(s"New $id is terminal ($condition) on agent $agentId during app $pathId restart (reservation: ${instance.reservation}) and the goal ($goal) is *NOT* Running! This means that someone is interfering with current deployment!")
+  // Kill next old instance
+  def killing: Receive = {
+    case KillImmediately(oldInstances) =>
+      logPrefixedInfo("killing")(s"Killing $oldInstances immediately.")
+      instances.valuesIterator
+        .filter { instance =>
+          instance.runSpecVersion < runSpec.version && instance.state.goal == Goal.Running
+        }
+        .take(oldInstances)
+        .foldLeft(Future.successful(Seq.empty[Instance.Id])) { (acc, nextDoomed) =>
```

Comment:
What do you think about using `Seq` instead of `Vector`?

Reply:
Sounds good. We probably should go with `Vector` anyways.